### PR TITLE
[Slim-LM] Enable Group Quant

### PR DESCRIPTION
**Pull request by zxybazh**
This PR introduces Group Quant via the new Slim-LM interface.

### CODE REVIEW

**Comment by junrushao** [comment]
CC: @cyx-6 please review

**Review by cyx-6 — APPROVED** [review]
LGTM!

File: python/mlc_chat/compiler/model/llama_quantization.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
+"""
+Quantization specs for Llama2 architecture.
+TODO: add docstring
```
Comment (junrushao) [inline python/mlc_chat/compiler/model/llama_quantization.py:3]: From my POV, the docstring looks pretty reasonable. Any specific ones in your mind?

File: python/mlc_chat/compiler/parameter/huggingface_loader.py [hunk 2]
```diff
@@ -66,12 +76,17 @@ def __init__(
...
+        quantize_param_map: Optional[QuantizeMapping]
+            The quantization mapping from MLC to quantized MLC parameters, default to None, which
+            means no quantization.
         """
         assert path.is_file()
         self.stats = Stats()
         self.extern_param_map = extern_param_map
         self.cached_files = {}
         self.torch_to_path = {}
+        self.quantize_param_map = quantize_param_map
```
Comment (junrushao) [inline python/mlc_chat/compiler/parameter/huggingface_loader.py:89]: nit: reorder to be consistent with type annotation

```suggestion
        assert path.is_file()
        self.stats = Stats()
        self.cached_files = {}
        self.torch_to_path = {}
        self.extern_param_map = extern_param_map
        self.quantize_param_map = quantize_param_map
```

File: python/mlc_chat/compiler/parameter/utils.py [hunk 3]
```diff
@@ -1,15 +1,51 @@
...
 
+if TYPE_CHECKING:
+    from tvm.runtime import NDArray
+
+    from ..parameter import QuantizeMapping
+
 logger = logging.getLogger(__name__)
 
 
+class ParamQuantizer:
```
Comment (junrushao) [inline python/mlc_chat/compiler/parameter/utils.py:19]: we unlikely need this helper class - seems its core logic is ~10 lines of code?

File: python/mlc_chat/compiler/parameter/huggingface_loader.py [hunk 4]
```diff
@@ -90,7 +105,21 @@ def load(self) -> Iterator[Tuple[str, NDArray]]:
         mlc_names = _loading_order(self.extern_param_map, self.torch_to_path)
         for mlc_name in tqdm(mlc_names):
             param = self._load_mlc_param(mlc_name)
-            yield mlc_name, param
+            if self.quantize_param_map:
```
Comment (junrushao) [inline python/mlc_chat/compiler/parameter/huggingface_loader.py:108]: handle the case that `param` is not in `self.quantize_param_map` - meaning no quantization

File: python/mlc_chat/compiler/quantization/group_quantizer.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
+"""A group quantizer for on the fly parameter quantization"""
+# pylint: disable=too-few-public-methods
+
+from typing import List, Tuple
+
+from tvm import te, tir
+
+from .quantization import QuantizeConfig
+
```
Comment (junrushao) [inline python/mlc_chat/compiler/quantization/group_quantizer.py:9]: define the dataclass `GroupQuantizeConfig` here

File: python/mlc_chat/compiler/quantization/group_quantizer.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+def te_quantize(
+    weight: te.Tensor, config: QuantizeConfig
+) -> Tuple[te.Tensor, te.Tensor, List[te.Tensor]]:
+    """Group quantization for weight tensor, defined in tensor expression."""
+    # pylint: disable=too-many-locals
+    assert len(weight.shape) == 2
+    n, m = weight.shape
```
Comment (junrushao) [inline python/mlc_chat/compiler/quantization/group_quantizer.py:17]: nit: use `k` to denote the reduction axis

```suggestion
    n, k = weight.shape
```

**Pull request merged**